lib/eslint/eslint-helpers.js

```diff
@@ -277,94 +256,168 @@ async function globSearch({
+                // no match, so we can skip this directory
+                if (matchedPatterns.length === 0) {
+                    return false;
+                }
+
+                const absolutePath = normalizeToPosix(path.resolve(basePath, entry.path));
+
+                /*
+                 * This part is a bit involved because we need to account for
+                 * the different ways that the patterns can match directories.
+                 * For each different way, we need to decide if we should look
+                 * for a config file or just use the default config.
+                 *
+                 * In general, we assume that the user wants to lint anything
+                 * that is passed directly. This is complicated because we start
+                 * the search from the base directory and work our way down. So,
+                 * we need to know if we are in a directory that was passed directly
+                 * or if we are in a directory that was matched by a pattern in
+                 * order to continue the traversal.
+                 *
+                 * Here are the cases:
+                 *
+                 * 1. A directory is passed directly (e.g., "subdir"). In this case, we
+                 * can assume that the user intends to lint this directory and we should
+                 * not look for a config file in the parent directory, because the only
+                 * reason to do that would be to ignore this directory (which we already
+                 * know we don't want to do). Instead, we use the default config until we
+                 * get to the directory that was passed, at which point we start looking
+                 * for config files again.
+                 *
+                 * 2. A dot (".") or star ("*"). In this case, we want to read
+                 * the config file in the current directory because the user is
+                 * explicitly asking to lint the current directory. Note that "."
+                 * will traverse into subdirectories while "*" will not.
+                 *
+                 * 3. A directory is passed in the form of "subdir/subsubdir".
+                 * In this case, we don't want to look for a config file in the
+                 * parent directory ("subdir"). We can skip looking for a config
+                 * file until `entry.depth` is greater than 1 because there's no
+                 * way that the pattern can match `entry.path` yet.
+                 *
+                 * 4. A directory glob pattern is passed (e.g., "subd*"). We want
+                 * this case to act like case 2 because it's unclear whether or not
+                 * any particular directory is meant to be traversed.
+                 *
+                 * 5. A recursive glob pattern is passed (e.g., "**"). We want this
+                 * case to act like case 2.
+                 */
+                const shouldLoadConfigFile = matchedPatterns.some(pattern => {
```

Comment:
> Did you maybe consider the approach of always grouping patterns by their glob parent in `findFiles()`?

I didn't understand what you meant by this, so I didn't consider it. If you'd like to explain more, or maybe write up an example, I'm happy to.

I do think we're starting to get into edge cases, though, and it might be better to merge the PR and then tackle edge cases individually. That's part of the benefit of having this behind a flag.

Reply:
I was referring to this code:

https://github.com/eslint/eslint/blob/b0df2e371fa8fd097cb31d914181a54df58e9e7e/lib/eslint/eslint-helpers.js#L603-L642

Patterns inside `cwd` are in the same group, while the rest are grouped by `globParent(pattern)`. If we always group patterns by `globParent(pattern)`, then we'll have groups such that for all patterns in the group the same configs apply.

In the example from my comment above, `src/*.js` and `**/*.cjs` would be in different groups.

Reply:
Ah! Thank you for explaining. That makes sense. Let me give that a try.

Reply:
That does work for a lot of the cases, but for `npx eslint src`, it now throws an error saying it's unmatched. So, it seems like we'll still need additional logic. Just three failing tests, though, so it's close.

Reply:
Okay, I think this piece is working now. :tada: